kythe/go/util/riegeli/reader.go

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package riegeli
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+func (r *Reader) nextRecord() ([]byte, error) {
+	if err := r.readNextRecordChunk(); err != nil {
+		return nil, err
+	}
+
+	size, err := binary.ReadUvarint(r.sizesReader)
+	if err != nil {
+		return nil, fmt.Errorf("error reading record size: %v", err)
+	}
+	val := make([]byte, size)
+	if _, err := io.ReadFull(r.valsReader, val); err != nil {
+		return nil, fmt.Errorf("error reading record value: %v", err)
+	}
+	r.numRecords--
+	return val, nil
+}
+
+func (r *Reader) readNextRecordChunk() error {
+	for r.valsReader == nil || r.numRecords == 0 {
+		c, err := r.r.Next()
+		if err != nil {
+			return err
+		} else if c.Header.ChunkType != recordChunkType {
+			// TODO(schroederc): support non-record chunk types (especially RecordsMetadata)
+			continue
+		}
+
+		rec, err := decodeRecordChunk(c)
+		if err != nil {
+			return fmt.Errorf("error decoding record chunk: %v", err)
+		} else if rec.CompressionType != noCompression {
+			// TODO(schroederc): compression
+			return fmt.Errorf("unsupported compression type: %s", []byte{byte(rec.CompressionType)})
+		} else if uint64(len(rec.CompressedValues)) != c.Header.DecodedDataSize {
+			return fmt.Errorf("bad DecodedDataSize: %d vs %d", len(rec.CompressedValues), c.Header.DecodedDataSize)
+		}
+		r.numRecords = c.Header.NumRecords
+		r.sizesReader = bytes.NewReader(rec.CompressedSizes)
+		r.valsReader = bytes.NewReader(rec.CompressedValues)
+	}
+	return nil
+}
+
+type blockReader struct {
+	r   io.Reader
+	buf *bytes.Reader
+}
+
+// Read implements the io.Reader interface by skipping over the interleaven
```

Comment:
This seems a slightly odd interface to implement. What's the use case for reading the data stream without the record boundaries?

Reply:
It's without the block boundaries, not the chunk headers.  The block headers are unnecessary except for verifying their hash and seeking to a particular position.  I added the hash check and the seeking will happen in the future.